0015-estimator-interface.md

```diff
@@ -0,0 +1,500 @@
+is as follows:
+
+```python
+T = TypeVar("T", bound=Job)
+
+BindingsArrayLike = Union[
+    BindingsArray, 
+    NDArray, 
+    Dict[Parameter | Tuple[Parameter], NDArray]
+]
+
+ObservablesArrayLike = Union[
+    ObservablesArray,
+    Iterable[str | Pauli | SparsePauliOp]
+]
+
+EstimatorTaskLike = Union[
+    EstimatorTask,
+    Tuple[QuantumCircuit, ObservablesArrayLike, BindingsArrayLike]
+]
+
+class EstimatorBase(ABC, Generic[T]):
+    ...
+    def run(self, tasks: EstimatorTaskLike | Iterable[EstimatorTaskLike], **options) -> T:
+        pass
+```
+
+### Migration Examples <a name="migration-examples"></a>
+
+Technical details of migration are discussed in the [Migration Path](#migration-path) section, but it is perhaps worthwhile to show some examples that demonstrate that, despite the new array and broadcasting abilities, the interface is not overly complicated for simple examples.
+
+One circuit, many observables:
+
+```python
+    # old API
+    estimator.run([circuit] * 4, observables=["ZIII", "IZII", "IIZI", "IIIZ"]).result()
+    >> EstimatorResult(np.array([0.1, 0.2, 0.3, 0.4]), metadata={...})
+
+    # new API (we can use a single task)
+    estimator.run((circuit, ["ZIII", "IZII", "IIZI", "IIIZ"])).result()
+    >> [ResultTask({"evs": np.array([0.1, 0.2, 0.3, 0.4]), "stds": <...>}, metadata={...})]
+```
+
+One circuit, one observable, many parameter value sets:
+
+```python
+    # old API
+    estimator.run([circuit] * 3, observables=["XYZI"] * 3, parameter_values=[[1,2,3], [3,4,5], [8,7,6]]).result()
+    >> EstimatorResult(np.array([0.1, 0.2, 0.3]), metadata={...})
+
+    # new API (we can use a single task)
+    estimator.run((circuit, ["XYZI"], [[1,2,3], [3,4,5], [8,7,6]])).result()
+    >> [ResultTask({"evs": np.array([0.1, 0.2, 0.3]), "stds": <...>}, metadata={...})]
+```
+
+Three different circuits, three different observables:
+
+```python
+    # old API
+    estimator.run([circuit1, circuit2, circuit3], observables=["ZIII", "IZII", "IIZI"]).result()
+    >> EstimatorResult(np.array([0.1, 0.2, 0.3]), metadata={...})
+
+    # new API (unique circuits always need their own tasks)
+    estimator.run([(circuit1, "ZIII"), (circuit2, "IZII"), (circuit3, "IIZI")]).result()
+    >> [
+    >>     ResultTask({"evs": np.array(0.1), "stds": <...>}, metadata={...}), 
+    >>     ResultTask({"evs": np.array(0.2), "stds": <...>}, metadata={...}), 
+    >>     ResultTask({"evs": np.array(0.3), "stds": <...>}, metadata={...})
+    >> ]
+```
+
+
+## Detailed Design <a name="detailed-design"></a>
+
+### Tasks <a name="tasks"></a>
+
+We propose the concept of a _task_, which we define as _a single circuit along with auxiliary data required to execute the circuit relative to the primitive in question_. This concept is general enough that it can be used for all primitive types, current and future, where we stress that what the “auxiliary data” is can vary between primitive types. 
```

Comment:
Something I mentioned in the meeting yesterday (even though the general consensus was that it is beyond the scope of this RFC):

This RFC does not mention that concept of "shots" anywhere.  If multiple circuits are submitted as multiple tasks to the same job, then they currently must have the same `shots` count (as specified in the `run_options`).  However, there are use cases (such as circuit cutting) where the results of some circuits contribute more weight to the final result than others, and so it would be desirable to run those circuits/tasks with proportional effort, some with more shots than the others.  Even after this RFC, the only way to accomplish this will be to submit the circuits in separate jobs, but this would remove the benefits of [pipelining](https://en.wikipedia.org/wiki/Protocol_pipelining) them all into the same remote procedure call.

In some uses cases, it might even be beneficial for the shots count to be adjustable on a per-parameter basis, not just a per-task basis.

In response to my comment, somebody mentioned that it might be even better for the estimator to be controlled by a desired error threshold than a shots count.  But even in that case, it would be nice to apply this at a more fine level of granularity than per-job.

Reply:
I expect that this would be possible as an additive change, either by allowing another field to `EstimatorTask` that accepts precision, or in the `ObservablesArray`. Precision could possibly be array-based if you needed different precision for different estimates.